src/core/java/xyz/wagyourtail/jsmacros/core/service/EventService.java

```diff
@@ -20,14 +26,89 @@ public class EventService extends BaseEvent {
 
     /**
      * when this service is stopped, this is run...
+     * @see EventService#unregisterOnStop(boolean, Registrable[])
      */
     @Nullable
     public MethodWrapper<Object, Object, Object, ?> stopListener;
+    /**
+     * @see EventService#unregisterOnStop(boolean, Registrable[])
+     * @since 1.9.1
+     */
+    @Nullable
+    public MethodWrapper<Object, Object, Object, ?> postStopListener;
+    protected boolean offEventsOnStop = false;
+    @Nullable
+    protected Registrable<?>[] registrableList = null;
 
     public EventService(String name) {
         this.serviceName = name;
     }
 
+    /**
+     * Setup unregister on stop. For example, {@code event.unregisterOnStop(false, d2d);} is
+     * the equivalent of {@code event.stopListener = JavaWrapper.methodToJava(() => d2d.unregister());}.<br>
+     * <br>
+     * If this is called multiple times, the previous ones would be discarded.<br>
+     * <br>
+     * The order of execution is run stopListener -> off events -> unregister stuff -> run postStopListener.<br>
+     * <br>
+     * Note that if the script stopped by itself (stopped without you stopping it in the gui), this won't take effect.
+     * @param offEvents whether the service manager should clear event listeners on stop.
+     *                  it's generally recommended to set to true, because callbacks cannot run after context closed.
+     * @param list the list of registrable, such as Draw2D, Draw3D and CommandBuilder.
+     * @since 1.9.1
+     */
+    public void unregisterOnStop(boolean offEvents, Registrable<?> ...list) {
+        offEventsOnStop = offEvents;
+        registrableList = list.length > 0 ? list : null;
+
+        // TODO: prevent the script from stopping itself (idk how to do it - aMelonRind)
+        // example code of script stopping itself:
+        // JsMacros.assertEvent(event, 'Service');
+        // const d2d = Hud.createDraw2D().register();
+        // d2d.addText('Hello World', 200, 200, 0xFFFFFF, true);
+        // event.unregisterOnStop(true, d2d);
+    }
+
+    /**
+     * don't touch this... although it's nothing much dangerous.
+     * @since 1.9.1
+     */
+    @DocletIgnore
+    public void _onStop(BaseScriptContext<?> ctx, Consumer<Throwable> errorConsumer) {
```

Comment:
would this function be able to be in ServiceManager maybe. would be cleaner for the api I think.

Reply:
I was thinking about not exposing the `offEventsOnStop` and `registrableList` fields. now I reconsidered, it's probably not that bad exposing them. sure.

Reply:
maybe I can create static WeakMap<EventService, ?> on ServiceManager to store some value that we don't want to expose on the api..